### ISSUE TITLE
Configurable UID/GID AAD attribute and/or improve mapping

### ISSUE THREAD

**Issue by nsballmann** [ubuntu/aad-auth#208]
The [UID](https://github.com/ubuntu/aad-auth/blob/b401372f346fa928ce96186b0bffd3a9cd24cffc/internal/cache/cache.go#L339) (also used as [GID](https://github.com/ubuntu/aad-auth/blob/b401372f346fa928ce96186b0bffd3a9cd24cffc/internal/cache/cache.go#L352)) seems to be [device locally generated](https://github.com/ubuntu/aad-auth/blob/b401372f346fa928ce96186b0bffd3a9cd24cffc/internal/cache/cache.go#L408) based on the username and some heuristics to avoid local collisions.

As already stated in #207 in our company the UPN can change. Therefore this highly likely yields also new UID/GID values. If the username would be configurable to a more stable user object attribute like proposed in #207 this would also result in more stable UID/GID values.

Can the attribute to be taken as base for that computation/generation be configurable as well? For example the AAD object ID is guarantied to be unique and stable throughout the AAD and therefore way better suited IMHO.

It would be even better if the mapping computation would be a bi-directional one to one mapping. Then `aad-auth` can guaranty not have UID collisions with all AAD users on all (AAD) devices. This would enable scenarios like one NFS share for all AAD users. And audits would be way easier because there is an actual connection back from the UID towards the AAD user object.

Best case would probably be, if the AAD would actually provide an user object integer attribute for UID guarantied to be unique in the scope of the AAD. AFAIK there is no such thing, but it can probably already be added with extension attributes. Therefore a configurable option to disable a mapping (or the current heuristics) and just taking a configurable attribute value as UID would be an ideal option.

**Comment by didrocks** [comment]
You are right, and the 2 features requests are definitively related. Due to the same reason (we don’t have access to more user attributes in the first iteration), we should evolve, while keeping backward compability, the UID (and thus GID) generation strategy.
